### PR TITLE
Run2-TB30 Add possibility of storing energy deposit in passive layers

### PR DESCRIPTION
**Pull request by bsunanda**
Though written for TB application it is general enough for other applications

### CODE REVIEW

**Comment by bsunanda** [comment]
@cmsbuild Please test

File: SimG4CMS/HGCalTestBeam/plugins/HGCPassive.cc [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+  topPV_ = getTopPV();
+  if (topPV_ == 0) {
+    edm::LogWarning("HGCPassive") << "Cannot find top level volume\n";
+  } else {
+    init_ = true;
+    const G4LogicalVolumeStore * lvs = G4LogicalVolumeStore::GetInstance();
+    std::vector<G4LogicalVolume *>::const_iterator lvcite;
+
+    for (lvcite = lvs->begin(); lvcite != lvs->end(); lvcite++) {
```
Comment (kpedro88) [inline SimG4CMS/HGCalTestBeam/plugins/HGCPassive.cc]: Use range-based for loop (delete unnecessary `lvcite` iter above)

File: SimG4CMS/HGCalTestBeam/plugins/HGCPassive.cc [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+      findLV(*lvcite);
+    } 
+
+#ifdef EDM_ML_DEBUG
+    std::cout << "HGCPassive::Finds " << mapLV_.size() << " logical volumes\n";
+    std::map<G4LogicalVolume*,std::pair<unsigned int, std::string> >::iterator itr;
+    unsigned int k(0);
+    for (itr = mapLV_.begin(); itr != mapLV_.end(); ++itr, ++k)
```
Comment (kpedro88) [inline SimG4CMS/HGCalTestBeam/plugins/HGCPassive.cc]: use range-based for loop (delete unnecessary map itr above)

File: SimG4CMS/HGCalTestBeam/plugins/HGCPassive.cc [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+void HGCPassive::update(const G4Step * aStep) {
+
+  if (aStep != NULL) {
+
+    G4VSensitiveDetector* curSD = aStep->GetPreStepPoint()->GetSensitiveDetector();
+    if (curSD==NULL) {
+      
+      G4TouchableHistory* touchable = (G4TouchableHistory*)aStep->GetPreStepPoint()->GetTouchable();
+      G4LogicalVolume* plv = (G4LogicalVolume*)touchable->GetVolume()->GetLogicalVolume();
+      std::map<G4LogicalVolume*,std::pair<unsigned int,std::string>>::iterator it = (init_) ? mapLV_.find(plv) : findLV(plv);
```
Comment (kpedro88) [inline SimG4CMS/HGCalTestBeam/plugins/HGCPassive.cc]: just use `auto it`

**Comment by civanch** [comment]
@bsunanda , where from the list of Logical Volumes may be obtained? How to be sure that the list is complete? I understand that it is possible for new HGCal TB due to relatively regular structure.

More basic question: you create a special hit for this type of study, which includes PhysicalVolume name, why you need extra map inside this class connected which LogicalVolumes? The map may be huge. I would think that one should do one or another.

File: SimG4CMS/HGCalTestBeam/plugins/HGCPassive.cc [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+      if (it != mapLV_.end()) {
+	unsigned int copy = (unsigned int)(touchable->GetReplicaNumber(0) + 
+					   1000*touchable->GetReplicaNumber(1));
+	std::pair<G4LogicalVolume*,unsigned int> key(plv,copy);
+	std::map<std::pair<G4LogicalVolume*,unsigned int>,std::pair<double,double>>::iterator itr = store_.find(key);
```
Comment (kpedro88) [inline SimG4CMS/HGCalTestBeam/plugins/HGCPassive.cc]: just use `auto itr`

File: SimG4CMS/HGCalTestBeam/plugins/HGCPassive.cc [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+
+
+//================================================================ End of EVENT
+
+void HGCPassive::endOfEvent(edm::PassiveHitContainer& hgcPH, unsigned int k) {
+#ifdef EDM_ML_DEBUG
+  unsigned int kount(0);
+#endif
+  std::map<std::pair<G4LogicalVolume*,unsigned int>,std::pair<double,double>>::iterator itr;
+  for (itr = store_.begin(); itr != store_.end(); ++itr) {
```
Comment (kpedro88) [inline SimG4CMS/HGCalTestBeam/plugins/HGCPassive.cc]: use range-based for loop (delete unnecessary itr above)

File: SimG4CMS/HGCalTestBeam/plugins/HGCPassive.cc [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+    G4LogicalVolume* lv = (itr->first).first;
+    std::map<G4LogicalVolume*,std::pair<unsigned int,std::string>>::iterator it = mapLV_.find(lv);
```
Comment (kpedro88) [inline SimG4CMS/HGCalTestBeam/plugins/HGCPassive.cc]: just use `auto it`

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by bsunanda** [comment]
@civanch This is more or less what the TB people wanted. It will be used only for special studies not for regular analysis. I just checked the size of the map for 2017 TB application - it is not big at all. # of logical volumes is smaller than that for physical volumes or touchables.

File: SimG4CMS/HGCalTestBeam/plugins/HGCPassive.cc [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+
+HGCPassive::HGCPassive(const edm::ParameterSet &p) : count_(0), init_(false) {
+
+  edm::ParameterSet m_Passive = p.getParameter<edm::ParameterSet>("HGCPassive");
+  LVNames_  = m_Passive.getUntrackedParameter<std::vector<std::string> >("LVNames");
+
+#ifdef EDM_ML_DEBUG
+  unsigned int k(0);
+#endif
+  for (auto name : LVNames_) {
```
Comment (kpedro88) [inline SimG4CMS/HGCalTestBeam/plugins/HGCPassive.cc]: use `const auto&`

File: SimG4CMS/HGCalTestBeam/plugins/HGCPassive.cc [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+    init_ = true;
+    const G4LogicalVolumeStore * lvs = G4LogicalVolumeStore::GetInstance();
+    for (auto lvcite : *lvs) {
+      findLV(lvcite);
+    } 
+
+#ifdef EDM_ML_DEBUG
+    std::cout << "HGCPassive::Finds " << mapLV_.size() << " logical volumes\n";
+    unsigned int k(0);
+    for (auto lvs : mapLV_) {
```
Comment (kpedro88) [inline SimG4CMS/HGCalTestBeam/plugins/HGCPassive.cc]: use `const auto&`

File: SimG4CMS/HGCalTestBeam/plugins/HGCPassive.cc [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+}//end update aStep
+
+
+//================================================================ End of EVENT
+
+void HGCPassive::endOfEvent(edm::PassiveHitContainer& hgcPH, unsigned int k) {
+#ifdef EDM_ML_DEBUG
+  unsigned int kount(0);
+#endif
+  for (auto element : store_) {
```
Comment (kpedro88) [inline SimG4CMS/HGCalTestBeam/plugins/HGCPassive.cc]: use `const auto&`

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by kpedro88** [comment]
+1

**Comment by civanch** [comment]
+1

**Comment by davidlange6** [comment]
+1

**Pull request merged**